Refactor TypeExtractor

gselzer opened the pull request:
This PR refactors the `TypeExtractor` interface, providing much more power in the ability to determine the type of an `Object`.

The main use case driving this refactor is the ability to control the reification of types like `net.imagej.Dataset` - types that have no type variables, who we'd like to know the type variables for. This is vital for using ImageJ Ops2 in SciJava scripts, because we need to be able to reify `Dataset`s into a `Img<T>`.

Still need to fix the ImageJ Ops2 tests - probably a bug in one of the converted `TypeExtractor`s.

gselzer committed:
Refactor TypeExtractor

ctrueden commented on imagej/imagej-ops2/src/main/java/net/imagej/ops2/filter/gauss/Gaussians.java:
```diff
@@ -71,11 +70,11 @@ public class Gaussians {
-	public static <T extends NumericType<T> & NativeType<T>> void defaultGaussRAI(
-		final RandomAccessibleInterval<T> input, //
+	public static <S extends NumericType<S>, T extends NumericType<T>> void defaultGaussRAI(
+		final RandomAccessibleInterval<S> input, //
```
How about using `EI` for element type of input image, and something else consistent... `EB`?... for element type of `OutOfBoundsFactory`? We use `EI` for element type of input images and `EO` for element type of output images in ImageJ Ops v1 in a few places, and I'd like to see more consistency here if possible, to reduce possible confusion.

gselzer replied:
A quick grep search shows no instances of `EI`/`EB`/`EO` in the incubator, so I'd prefer to not introduce it.

I used `S` and `T` in this class because that's what the underlying `SeparableSymmetricConvolution` uses, but we could change it to `I`/`O`!

gselzer replied:
Renamed in 035e1c1

ctrueden commented on imagej/imagej-ops2/src/main/java/net/imagej/ops2/types/ImgFactoryTypeExtractor.java:
```diff
@@ -48,25 +49,21 @@
-
-		return t.reify(o.type());
+	@Override public double getPriority() {
+		return Priority.NORMAL;
```
Isn't `NORMAL` already the default? If not, shouldn't it be?

gselzer replied:
Yeah, it should be. Fixed in 877d83f

ctrueden commented on imagej/imagej-ops2/src/main/java/net/imagej/ops2/types/ImgFactoryTypeExtractor.java:
```diff
@@ -48,25 +49,21 @@
+			throw new IllegalArgumentException(this + " cannot reify " + object);
+		ImgFactory<?> imgFactory = (ImgFactory<?>) object;
+		Type componentType = r.reify(imgFactory.type());
+		return TypeTools.raiseParametersToClass(object.getClass(), ImgFactory.class, new Type[] {componentType});
```
In general, I love the generalization of the `reify` method. But I dislike this boilerplate `TypeTools.raiseParametersToClass` thing that is now needed in the majority of `TypeExtractor` implementations. Surely we can make the code more DRY somehow?

gselzer replied:
Made an abstract superclass in 4e897c7c

ctrueden commented on scijava/scijava-types/src/main/java/org/scijava/types/TypeTools.java:
```diff
@@ -302,6 +304,20 @@ else if (paramComponent instanceof TypeVariable) {
 		return true;
 	}
 
+	public static Type raiseParametersToClass(Class<?> cls, final Class<?> superCls, final Type[] superClsTypeVars) {
```
The `superClsTypeVars` could be a `Type...` instead, to make calling code a little more readable.

I would also rename this method to something less bulky, especially since it is called by nearly every `TypeExtractor` implementations as its final step.

gselzer replied:
Renamed in 035e1c1

ctrueden commented on scijava/scijava-types/src/main/java/org/scijava/types/extractors/ParameterizedTypeExtractor.java:
```diff
@@ -0,0 +1,198 @@
+			}
+		}
+
+		// fill in any remaining unresolved type parameters with wildcards
```
This comment style differs from earlier comments. Stick with one style.

gselzer replied:
I'm not fully sure what you meant, but I tried to fix it in 035e1c1

ctrueden replied:
I meant:
```
// fill in any remaining unresolved type parameters with wildcards
```
vs.
```
// Fill in any remaining unresolved type parameters with wildcards.
```

So, whether to write in complete sentences, or only small case, basically.

For many years I was a small-case no-period guy, but I kind of came around to the idea that reading comments should be like reading anything else written by someone: so make it complete sentences in most situations.

But that's only my personal preference. The main thing is just to be consistent—and I noticed that we had some of the former and others of the latter.

It's a nitpick for sure, though!~

ctrueden reviewed (commented):
Nice progress! I made some comments.

gselzer committed:
Create TypeExtractor.getPriority

This commit, in addition to removing some duplicate code, moves around
some priorities to make a little more sense

gselzer committed:
Create abstract class for subType parameterization

gselzer committed:
Various minor fixes

gselzer committed:
Create module imagej2-legacy

This module will house components integrating SciJava3 with ImageJ2

gselzer committed:
Update META-INFS

Not sure if this is needed for running in Fiji

hinerm committed:
Update LICENSE.txt copyright years

hinerm committed:
Remove unused import

hinerm committed:
Fix valueFactory reference
